modules/bidViewabilityIO.js

```diff
@@ -0,0 +1,85 @@
+export const IAB_VIEWABLE_DISPLAY_LARGE_THRESHOLD = 0.3;
+
+const CLIENT_SUPPORTS_IO = window.IntersectionObserver && window.IntersectionObserverEntry && window.IntersectionObserverEntry.prototype &&
+    'intersectionRatio' in window.IntersectionObserverEntry.prototype;
+
+const supportedMediaTypes = [
+  'banner'
+];
+
+export let isSupportedMediaType = (bid) => {
+  return supportedMediaTypes.includes(bid.mediaType);
+}
+
+// returns options for the iO that detects if the ad is viewable
+export let getViewableOptions = (bid) => {
+  if (bid.mediaType === 'banner') {
+    return {
+      root: null,
+      rootMargin: '0px',
+      threshold: bid.width * bid.height > IAB_VIEWABLE_DISPLAY_LARGE_PX ? IAB_VIEWABLE_DISPLAY_LARGE_THRESHOLD : IAB_VIEWABLE_DISPLAY_THRESHOLD
+    }
+  }
+}
+
+// markViewed returns a function what will be executed when an ad satisifes the viewable iO
+export let markViewed = (bid, entry, observer) => {
+  return () => {
+    observer.unobserve(entry.target);
+    events.emit(EVENTS.BID_VIEWABLE, bid);
+    logMessage(`id: ${entry.target.getAttribute('id')} code: ${bid.adUnitCode} was viewed`);
+  }
+}
+
+// viewCallbackFactory creates the callback used by the viewable IntersectionObserver.
+// When an ad comes into view, it sets a timeout for a function to be executed
+// when that ad would be considered viewed per the IAB specs. The bid that was rendered
+// is passed into the factory, so it can pass it into markViewed, so that it can be included
+// in the BID_VIEWABLE event data. If the ad leaves view before the timer goes off, the setTimeout
+// is cancelled, an the bid will not be marked as viewed. There's probably some kind of race-ish
+// thing going on between IO and setTimeout but this isn't going to be perfect, it's just going to
+// be pretty good.
+export let viewCallbackFactory = (bid) => {
+  return (entries, observer) => {
+    entries.forEach(entry => {
+      if (entry.isIntersecting) {
+        logMessage(`viewable timer starting for id: ${entry.target.getAttribute('id')} code: ${bid.adUnitCode}`);
+        entry.target.view_tracker = setTimeout(markViewed(bid, entry, observer), IAB_VIEWABLE_DISPLAY_TIME);
+      } else {
+        logMessage(`id: ${entry.target.getAttribute('id')} code: ${bid.adUnitCode} is out of view`);
+        if (entry.target.view_tracker) {
+          clearTimeout(entry.target.view_tracker);
+          logMessage(`viewable timer stopped for id: ${entry.target.getAttribute('id')} code: ${bid.adUnitCode}`);
+        }
+      }
+    });
+  };
+};
+
+export let init = () => {
+  events.on(EVENTS.AD_RENDER_SUCCEEDED, ({doc, bid, id}) => {
+    // read the config for the module
+    const globalModuleConfig = config.getConfig(MODULE_NAME) || {};
+    // do nothing if module-config.enabled is not set to true
+    // this way we are adding a way for bidders to know (using pbjs.getConfig('bidViewability').enabled === true) whether this module is added in build and is enabled
```

Comment:
I think it would be better to never register for this event until the user enables the config.

So your init function could subscribe to setConfig and then once it is enabled, it then will add the event listener.

```js
config.getConfig(MODULE_NAME, conf => {
   // if conf enabled then 
   events.on(EVENTS.AD_RENDER_SUCCEEDED, ({doc, bid, id}) => {
});
```

That way the event callback is not executing every single time this module is included.

Only when it is actually turned on.

I would also add the `CLIENT_SUPPORTS_IO` check inside of that config subscribe as well since no need to run it when it is not supported

Reply:
i just pushed a change that doesn't do the subscribe thing, then actually read all these words and found out that this subscribe thing is a thing.  I'll have a look it doing it as you describe this afternoon probably.

Reply:
Ok cool.

So basically if you do the following

```js
config.getConfig(MODULE_NAME, CALLBACK_FUNCTION);
```

Prebid will execute your `CALLBACK_FUNCTION` each time `pbjs.setConfig` is called with your `MODULE_NAME`

So for your use case, we only want to register the event listener if the user / pub runs pbjs.setConfig and turns on bidViewability.

so your callback can do something like:

```js
config.getConfig(MODULE_NAME, conf => {
   const  globalModuleConfig = conf.bidViewabilityIO;
   if (globalModuleConfig[CONFIG_ENABLED] && CLIENT_SUPPORTS_IO) {
      // register your event code
      events.on(EVENTS.AD_RENDER_SUCCEEDED, ({doc, bid, id}) => {
   }
});
```